### DISCUSSION TITLE
Conflicts between millis() and tone()

### DISCUSSION THREAD

**Discussion by kr4fty** [stm32duino/Arduino_Core_STM32#2498] [General]
Hi everyone, I'm working with a stm32F103 mcu. It turns out that I use millis() to refresh the reading of data through analogRead. The thing is that I use reading from a button, which would make me activate a process and also produce a tone, through tone(). I notice that when I use millis it stops at that moment and continues once the tone() function has finished, as if tone did not take any time.

The problem comes because both millis() and tone() use the same timer?

my code:

```
#include <Arduino.h>

#define LED LED_BUILTIN

void setup()
{
    pinMode(LED, OUTPUT);
    digitalWrite(LED, LOW);

    pinMode(BUTTON_PIN, INPUT_PULLUP);
    
    Serial.begin(9600);
}

bool isButtonClicked;
unsigned long start, end, x;
void loop()
{
    if(!digitalRead(BUTTON_PIN)){
        delay(25);
        if(digitalRead(BUTTON_PIN))
            isButtonClicked = true;
    }
    if (isButtonClicked){
        isButtonClicked = false;
      
        start = millis();
        tone(BUZZER_PIN, 4000, 40);
        end = millis();
        digitalWrite(LED, LOW);

        Serial.print(end-start);
   }
}
```

my output:

```
0
0
0
0
```

**Comment by fpistm** [comment]
Hi @kr4fty 
There is no issue from my point of view. 
`HardwareTimer` is used for the `Tone` so you used Hardware ressources.
While `millis` uses the `systick`.

So when you do:
```
        start = millis();
        tone(BUZZER_PIN, 4000, 40);
        end = millis();
```
`Tone` simply configure the `HardwareTimer` and exit immediately. So having 0 is coherent.
